File: arai60/number-of-islands/step3.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        LAND = '1'
        WATER = '0'
        n_row = len(grid)
        n_col = len(grid[0])
        visited = [[False] * n_col for _ in range(n_row)]

        def explore_island(node: tuple) -> None:
            current_row, current_col = node
            visited[current_row][current_col] = True
            for next_node in get_neighbors(node):
                next_row, next_col = next_node
                if not visited[next_row][next_col]:
                    explore_island(next_node)

        def get_neighbors(node: tuple) -> list:
            drow = (0, 1, 0, -1)
            dcol = (1, 0, -1, 0)
            neighbors = []
            for direction in range(4):
                next_row = node[0] + drow[direction]
                next_col = node[1] + dcol[direction]
                if 0 <= next_row < n_row and 0 <= next_col < n_col and grid[next_row][next_col] == LAND:
                    neighbors.append((next_row, next_col))

            return neighbors

        total = 0
        for i in range(n_row):
            for j in range(n_col):
                if grid[i][j] == LAND and not visited[i][j]:
                    explore_island((i, j))
                    total += 1
        return total

File: arai60/number-of-connected-components-in-an-undirected-graph/step3.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adjacency_list(n);
        vector<bool> visited(n, false);

        for (vector<int> edge : edges) {
            adjacency_list[edge[0]].push_back(edge[1]);
            adjacency_list[edge[1]].push_back(edge[0]);
        }

        int total = 0;
        for (int i = 0; i < n; i++) {
            if (visited[i]) continue;
            visitAllConnectedNodes(i, adjacency_list, visited);
            total++;
        }
        return total;
    }

private:
    void visitAllConnectedNodes(int node, vector<vector<int>>& adjacency_list, vector<bool>& visited) {
        visited[node] = true;
        for (int next_node : adjacency_list[node]) {
            if (visited[next_node]) continue;
            visitAllConnectedNodes(next_node, adjacency_list, visited);
        }
    }
};
